src/L10NSharp/XLiffUtils/XLiffBody.cs

```diff
@@ -29,35 +32,93 @@ public class XLiffBody
 		// This is used when translation unit IDs are not found in the file (which seems to be
 		// the case with Lingobit XLiff files).
 		private int _transUnitId;
-		private bool _idsVerified;
-		private List<XLiffTransUnit> _transUnits = new List<XLiffTransUnit>();
+		static SpinLock _transUnitIdLock;
+
+		private ConcurrentDictionary<string, XLiffTransUnit> _transUnitDict =
+			new ConcurrentDictionary<string, XLiffTransUnit>();
+
+		private object mutex = new object(); // lock for accessing non-concurrent variables.
 		private int _translatedCount = -1;
 		private int _approvedCount = -1;
 
+		public class ListWrapper : IEnumerable<XLiffTransUnit>
+		{
+			private IEnumerable<XLiffTransUnit> _list;
+			private XLiffBody _body;
+
+			public ListWrapper(IEnumerable<XLiffTransUnit> startWith, XLiffBody body)
+			{
+				_list = startWith;
+				_body = body;
+			}
+			public IEnumerator<XLiffTransUnit> GetEnumerator()
+			{
+				return _list.GetEnumerator();
+			}
+
+			IEnumerator IEnumerable.GetEnumerator()
+			{
+				return GetEnumerator();
+			}
+
+			public void Add(XLiffTransUnit item)
+			{
+				_body.AddTransUnitRaw(item);
+			}
+		}
+
 		#region Properties
 		/// ------------------------------------------------------------------------------------
 		/// <summary>
 		/// Gets the list of translation units in the file.
+		/// Note, a new list is generated each time, so this can be inefficient. Consider using
+		/// TransUnitsUnordered if you don't absolutely need a list. This method is retained
+		/// mainly for backwards compatibility.
```

Comment:
Since the behavior is subtly changing, maybe trying to preserve "backwards compatibility" is a bad thing.